packages/alfa-table/src/table.ts

```diff
@@ -224,29 +270,116 @@ export namespace Table {
               height: Math.max(this.height, this.height + rowGroup.height),
               width: Math.max(this.width, rowGroup.width),
               // merge in new cells
-              cells: this._cells.concat(...rowGroup.cells),
+              cells: this._cells.concat(rowGroup.cells),
               // add new group
-              rowGroups: [...this.rowGroups].concat(rowGroup.rowgroup),
+              rowGroups: List.from(this.rowGroups).append(rowGroup.rowgroup),
             });
           } else {
             return this;
           }
         });
     }
 
+    public hasDataCellCoveringArea(
+      x: number,
+      y: number,
+      w: number,
+      h: number
+    ): boolean {
+      return some(this.cells, (cell) => cell.isDataCoveringArea(x, y, w, h));
+    }
+
+    public addHeadersVariants(): Builder {
+      return this.update({
+        cells: List.from(
+          map(this.cells, (cell) =>
+            cell.addHeaderVariant(
+              this.hasDataCellCoveringArea.bind(this),
+              this.width,
+              this.height
+            )
+          )
+        ),
+      });
+    }
+
+    /**
+     * If principal cell is in a group, get all group headers that are in this group and above+lift of principal cell.
+     */
+    public getAboveLeftGroupHeaders(
+      kind: "row" | "column"
+    ): (principalCell: Cell.Builder) => Iterable<Cell.Builder> {
+      let anchor: "x" | "y",
+        groups: Iterable<Covering>,
```

Comment:
I'm not a huge fan of this `Covering` interface. Does the job, but feels a bit dirty…

Reply:
🤦 Of course 🙈